qutip/qobjevo.py

```diff
@@ -539,42 +543,39 @@ def __init__(self, Q_object=[], args={}, copy=True,
         if state0 is not None:
             self._dynamics_args_update(0., state0)
 
-    def _td_format_check_single(self, Q_object, tlist=None):
-        op_type = []
-
+    def _td_format_check(self, Q_object):
         if isinstance(Q_object, Qobj):
-            op_type = 0
-        elif isinstance(Q_object, (FunctionType,
-                                   BuiltinFunctionType, partial)):
-            op_type = 1
-        elif isinstance(Q_object, list):
-            if (len(Q_object) == 0):
-                op_type = -1
-            for op_k in Q_object:
-                if isinstance(op_k, Qobj):
-                    op_type.append(0)
-                elif isinstance(op_k, list):
-                    if not isinstance(op_k[0], Qobj):
-                        raise TypeError("Incorrect Q_object specification")
-                    elif len(op_k) == 2:
-                        if isinstance(op_k[1], Cubic_Spline):
-                            op_type.append(4)
-                        elif callable(op_k[1]):
-                            op_type.append(1)
-                        elif isinstance(op_k[1], str):
-                            op_type.append(2)
-                        elif isinstance(op_k[1], np.ndarray):
-                            if not isinstance(tlist, np.ndarray) or not \
-                                        len(op_k[1]) == len(tlist):
-                                raise TypeError("Time list does not match")
-                            op_type.append(3)
-                        else:
-                            raise TypeError("Incorrect Q_object specification")
-                    else:
-                        raise TypeError("Incorrect Q_object specification")
+            return 0
+        if isinstance(Q_object, (FunctionType, BuiltinFunctionType, partial)):
+            return 1
+        if isinstance(Q_object, list):
+            return [self._td_op_type(element) for element in Q_object] or -1
+        raise TypeError("Incorrect Q_object specification")
+
+    def _td_op_type(self, element):
+        if isinstance(element, Qobj):
+            return 0
+        try:
+            op, td = element
+        except (TypeError, ValueError) as exc:
+            raise TypeError("Incorrect Q_object specification") from exc
+        if (not isinstance(op, Qobj)) or isinstance(td, Qobj):
+            # Qobj is itself callable, so we need an extra check to make sure
+            # that we don't have a two-element list where both are Qobj.
+            raise TypeError("Incorrect Q_object specification")
+        if isinstance(td, Cubic_Spline):
+            out = 4
+        elif callable(td):
```

Comment:
We don't test the validity of the function in `__init__` because of the `dynamics_args`: passed `args` could not be usable before parsing them with `_args_checks`. Do you think that testing them if `_args_checks` did no work would be worth it?

Reply:
I'm not sure, to be honest - I didn't think about it at all here.  I only changed these parts of `QobjEvo` because I needed CodeClimate to be quiet while I made an unrelated change.  At least in the 4.x branch, I think what we've got is fine.  I think it'll be much easier to be rigorous in the 5.x branch with each term individually gaining its own `Coefficient`.

I didn't even want to make this change, since it's just one more thing to merge up into `dev.major`, but oh well...

I think that right now we shouldn't bother testing functions for validity - with the way dynamic arguments are handled in `QobjEvo` right now, it would probably involve some rather difficult logic to organise that, and I really wouldn't be convinced we'd get all the edge cases.  We can't catch every possible error a function could give anyway, so perhaps we should only attempt tests that are convenient for us to do?  I suppose we have to have some level of trust that the user will pass something usable.